src/relay/backend/manifest_lifetimes.cc

```diff
@@ -0,0 +1,367 @@
+  worklist.insert(worklist.end(), cfg.reverse_post_order.rbegin(), cfg.reverse_post_order.rend());
+
+  // See https://lambda.uta.edu/cse5317/notes/node40.html for an overview of the algorithm.
+  auto visitor = [&](const CFG::NodePtr n) {
+    VarSet old_in_n = a.live_in[n];
+    VarSet old_out_n = a.live_out[n];
+
+    a.live_in[n] = use_def.use.at(n);
+    for (const Var& v : a.live_out[n]) {
+      if (!v.same_as(use_def.def.at(n))) {
+        a.live_in[n].insert(v);
+      }
+    }
+
+    a.live_out[n] = VarSet();
+    for (const CFG::NodePtr& s : n->GetSucc()) {
+      a.live_out[n].insert(a.live_in[s].begin(), a.live_in[s].end());
+    }
+
+    if (SetEqual(old_in_n, a.live_in[n]) && SetEqual(old_out_n, a.live_out[n])) {
+      // No need to update the worklist.
+    } else {
+      // Add predecessor nodes back to worklist (no need to add successors, since each node's
+      // in/out sets are not dependent on its predecessors).
+      for (const CFG::NodePtr& p : n->GetPred()) {
+        worklist.push_back(p);
+      }
+    }
+  };
+
+  while (!worklist.empty()) {
+    const CFG::NodePtr n = worklist.front();
+    worklist.pop_front();
+    visitor(n);
+  }
+
+  return a;
+}
+
+Expr KillInserter::VisitExpr_(const LetNode* let_node) {
+  Expr expr = GetRef<Expr>(let_node);
+  LetList ll;
+
+  while (const LetNode* inner_let_node = expr.as<LetNode>()) {
+    ll.Push(inner_let_node->var, VisitExpr(inner_let_node->value));
+
+    ICHECK(!inner_let_node->value.as<VarNode>()) << "aliasing should have been eliminated.";
+    ICHECK(cfg_->let_map.count(expr)) << "all Let exprs should be mapped in the CFG";
+
+    const ControlFlowGraph::NodePtr n = cfg_->let_map.at(expr);
+
+    const VarSet& li = lva_->live_in.at(n);
+    const VarSet& lo = lva_->live_out.at(n);
+
+    // Killed vars = live in - live out.
+    VarSet kills;
+    for (const Var& v : li) {
+      if (!lo.count(v)) {
+        kills.insert(v);
+      }
+    }
+
+    for (const Var& v : kills) {
+      ll.Push(Call(Op::Get("memory.kill"), {v}));
+    }
+
+    expr = inner_let_node->body;
+  }
+
+  return ll.Get(VisitExpr(expr));
+}
+
+Expr AliasEliminator::VisitExpr_(const LetNode* let_node) {
+  Expr expr = GetRef<Expr>(let_node);
+  LetList ll;
+  std::vector<Var> aliased_vars;
+
+  while (const LetNode* inner_let_node = expr.as<LetNode>()) {
+    const Var& var = inner_let_node->var;
+    const Expr& val = inner_let_node->value;
+    bool aliased = false;
+    ICHECK(!alias_.count(var));
+
+    if (const VarNode* alias_of_n = AsIgnoringOnDevice<VarNode>(val)) {
+      alias_[var] = Downcast<Var>(VisitExpr_(alias_of_n));
+      aliased = true;
+    } else if (AsIgnoringOnDevice<CallNode>(val)) {
+      // Copying to the same device is aliasing.
+      // WARNING: this must be kept in sync with the VM compiler logic in
+      // src/relay/backend/vm/compiler.cc, line 541, in DeviceAwareVisitExpr_(const CallNode*).
```

Comment:
I feel a little weird about this move, since there's still some VM-specific logic in the stuff for manifesting lifetimes. The other stuff can probably be factored out though (CFG, and the basic analyses).

Reply:
Sure, I'll try to move some VM specific stuff back